### ISSUE TITLE
Improve linting for decorator functions

### ISSUE THREAD

**Issue by maximeflya** [jsh9/pydoclint#240]
The linting for decorators could be improved. Here is an example:

```python
from typing import Callable
from functools import wraps

def foo(func:Callable[[], None])->Callable[[], None]:
    """Print things decorator.

    :raises ValueError: Something went bad
    """
    @wraps(func)
    def wrapper()->None:
        print("a")
        func()
        raise ValueError("Oops")
    return wrapper

@foo
def bar():
    print("b")

bar()
```
This is how I would expect to have to document this decorator. But I get the following error:

```
4: DOC101: Function `foo`: Docstring contains fewer arguments than in function signature.
4: DOC103: Function `foo`: Docstring arguments are different from function arguments. (Or could be other formatting issues: https://jsh9.github.io/pydoclint/violation_codes.html#notes-on-doc103 ). Arguments in the function signature but not in the docstring: [func: Callable[[], None]].
4: DOC201: Function `foo` does not have a return section in docstring
4: DOC502: Function `foo` has a "Raises" section in the docstring, but there are not "raise" statements in the body
```

Another example with a decorator with an argument

```python 
from functools import wraps
from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")


def foo(nb: int) -> Callable[[Callable[P, R]], Callable[P, R]]:
    """The foo decorator.

    :param nb: a valid number
    :raises ValueError: some `nb`s are illegal
    :return: unchanged return value of wrapped function
    """

    def decorator(func: Callable[P, R]) -> Callable[P, R]:
        @wraps(func)
        def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            if nb == 2:
                raise ValueError()
            return func(*args, **kwargs)

        return wrapper

    return decorator


@foo(3)
def bar(val: str) -> int:
    return 3


print(bar("2"))
```

With this I get the following error:
```
8: DOC502: Function `foo` has a "Raises" section in the docstring, but there are not "raise" statements in the body
```
There is no return warning because the function returns the decorator function, but in reality, it is not checking for the correct return statement

**Idea to fix**

I don't know if it possible, but pydoclint could check if the function returns another function and in this case append the Exception and change the return to what the sub function returns.
I'm not sure how you would handle the `func` argument in the first case, because in the second case, we would like to document the decorator parameters. 

There is also the fact that we could have a function, which is not intended to be a decorator, that returns another callable. In this case, it might be good to have both the current linting and the updated behavior to be valid.

There might be other cases to consider.

This is based on how I would document these decorators but I might be doing things wrong.

Decorators are a bit niche but it would be nice to have good documentation for them :)

**Comment by jsh9** [comment]
Hi! Sorry for getting back to you so late!

I don't use decorators often, so I don't know what the best practice/convention is for documenting decorator functions. Could you find some examples in other open source projects? Thanks!

**Comment by aecorn** [comment]
Hi, I want to contribute an example from our codebase to this issue as an example of decorator usage and interaction with pydoclint.
(We just switched from darglint to pydoclint, thank you for this great project!)
It would be nice with a configuration option for "ignore certain errorcodes for functions decorated with these specific decorators" or similar.
Something like this perhaps:
```toml
specific-decorator-ignore-codes = {"wrap_derive": ["DOC101", "DOC103", "DOC201", "DOC203"]}
```

I dont know if this also would cover your usage of `@wraps` @maximeflya ?

Currently we are solving this by inling the #noqa: comments in our function definitions
This however is stripped by RUFF, so we need an ignore rule for RUFF too: `"src/nudb_use/variables/derive/*.py" = ["RUF100"]` - not very elegant, and the inlined comments will become quite repetitive.

In our code we wrap simple hardcoded functions with the decorator "wrap_derive", including a minimal docstring.
```python
@wrap_derive
def vg_kompetanse_nus(  # noqa: DOC101,DOC103,DOC201,DOC203
    df: pd.DataFrame,
) -> pd.Series:
    """Derive vg_kompetanse_nus from nus2000."""
    return _map_nus_variant(df, "vg_kompetanse_nus")

```

We then expand the docstring in the decorator-constructed wrapper, as this will be shared across all decorated functions.
```python
...
    wrapper.__name__ = basefunc.__name__
    docstring = basefunc.__doc__ or ""
    wrapper.__doc__ = f"""{docstring}

            Args:
                df: Dataframe that should contain prerequisites listed in {derived_from}.
                priority: 'old' keeps existing {name} values when present, 'new' prefers freshly derived values.

            Returns:
                pd.DataFrame: The dataframe with {name} added/updated when all prerequisites are available.
        """
    return wrapper
```
